### PR TITLE
CSHARP-3845: Deserialize Select of anonymous types using default values for missing fields.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

**Comment by rstam** [comment]
Evergreen patch build:

https://evergreen.mongodb.com/version/61afe46f306615750a434767

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewExpressionToAggregationExpressionTranslator.cs [hunk 1]
```diff
@@ -50,7 +50,8 @@ public static AggregationExpression Translate(TranslationContext context, NewExp
                 var fieldExpression = expression.Arguments[i];
                 var fieldTranslation = ExpressionToAggregationExpressionTranslator.Translate(context, fieldExpression);
                 var memberSerializer = fieldTranslation.Serializer ?? BsonSerializer.LookupSerializer(fieldExpression.Type);
-                classMap.MapProperty(member.Name).SetSerializer(memberSerializer);
+                var defaultValue = GetDefaultValue(memberSerializer.ValueType);
+                classMap.MapProperty(member.Name).SetSerializer(memberSerializer).SetDefaultValue(defaultValue);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NewExpressionToAggregationExpressionTranslator.cs:54]: This is really all there is to this... configure a default value for the member.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs [hunk 2]
```diff
@@ -29,6 +29,23 @@ protected void AssertStages(IEnumerable<BsonDocument> stages, IEnumerable<string
...
+            var documentsList = documents?.ToList();
+            if (documentsList != null && documentsList.Count == 0)
+            {
+                database.CreateCollection(collectionName);
+            }
+            else
+            {
+                collection.InsertMany(documentsList);
+            }
+        }
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs:47]: Given that `InsertMany` accepts an `IEnumerable<T>`, we don't need the intermediate `documentsList`. Consider the following instead. (I flipped if/else branches around so we don't have the awkward `!documents.Any()` construct:

```
protected void CreateCollection<TDocument>(IMongoCollection<TDocument> collection, IEnumerable<TDocument> documents = null)
{
    var database = collection.Database;
    var collectionName = collection.CollectionNamespace.CollectionName;
    database.DropCollection(collectionName);

    if (documents != null && documents.Any())
    {
        collection.InsertMany(documents);
    }
    else
    {
        database.CreateCollection(collectionName);
    }
}
```

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3845Tests.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            };
+            AssertStages(stages, expectedStages);
+
+            var results = queryable.ToList();
+            results.Count.Should().Be(2);
+            results[0].F.Should().Be(null);
+            results[0].G.Should().Be(0);
+            results[1].F.Should().Be("abc");
+            results[1].G.Should().Be(123);
+        }
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp3845Tests.cs:53]: We are assuming order of returned results without specifying a sort. We should specify a sort.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Nice simple solution. Some minor testing improvements and fixes requested.

**Review by rstam — COMMENTED** [review]
I don't object to any of your suggestions but I have questions/comments for you before making the changes.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs (hunk 2 above)
Comment (JamesKovacs) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs:47]: In production code, I am definitely concerned with possible multiple enumeration. But given this is test code and all code paths supply it with an array, I'm not worried about it.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Feedback on discussion points.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs [hunk 4]
```diff
@@ -29,6 +29,23 @@ protected void AssertStages(IEnumerable<BsonDocument> stages, IEnumerable<string
...
         }
 
+        protected void CreateCollection<TDocument>(IMongoCollection<TDocument> collection, IEnumerable<TDocument> documents = null)
+        {
+            var database = collection.Database;
+            var collectionName = collection.CollectionNamespace.CollectionName;
+            database.DropCollection(collectionName);
+
+            var documentsList = documents?.ToList();
+            if (documentsList != null && documentsList.Count == 0)
```
[the remark this answers is not in the document]
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs (hunk 2 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Linq3IntegrationTest.cs:46]: Done

But we should get in the habit of avoiding multiple enumeration... it's a bad practice. Being lazy in test code could lead to being lazy in production code.

It's a problem we were not aware of for a long time. There are multiple places in our production code where we inadvertently enumerate something multiple times.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request closed without merging**